File: scala/leet_code/402.scala
object Solution {

  /** This function removes k digits from the input number string to make the resulting number as small as possible.
    *
    * @param num The input number as a string. It should be a non-negative integer.
    * @param k   The number of digits to remove from the input number. It should be a non-negative integer.
    * @return A string representing the smallest possible number after removing k digits. If the resulting number is empty, return "0".
    * @example removeKdigits("1432219", 3) returns "1219".
    * @example removeKdigits("10200", 1) returns "200".
    * @example removeKdigits("10", 2) returns "0".
    */
  def removeKdigits(num: String, k: Int): String = {
    // Create a mutable stack to hold the digits of the number
    val stack = scala.collection.mutable.Stack[Char]()

    // Initialize the number of digits to remove
    var remove = k

    // Iterate over each character in the number string
    for (c <- num) {
      // While there are still digits to remove and the top of the stack is greater than the current digit, pop the stack
      while (remove > 0 && stack.nonEmpty && stack.top > c) {
        stack.pop()
        remove -= 1
      }
      // Push the current digit onto the stack
      stack.push(c)
    }

    // While there are still digits to remove, pop the stack
    while (remove > 0) {
      stack.pop()
      remove -= 1
    }

    // Create a new StringBuilder to hold the resulting number
    val sb = new StringBuilder()

    // While the stack is not empty, insert the top of the stack at the beginning of the StringBuilder
    while (stack.nonEmpty) {
      sb.insert(0, stack.pop())
    }

    // Convert the StringBuilder to a string and remove leading zeros
    val res = sb.toString().dropWhile(_ == '0')

    // If the resulting string is empty, return "0", otherwise return the resulting string
    if (res.isEmpty()) "0" else res
  }
}

File: php/leet_code/402.php
<?php

declare(strict_types=1);

/**
 * Class Solution
 */
class Solution
{
    /**
     * This function removes k digits from the number to make it smallest.
     *
     * @param string $num The number from which digits are to be removed.
     * @param int $k The number of digits to remove.
     * @return string The smallest possible number after removing k digits.
     */
    public function removeKdigits(string $num, int $k): string
    {
        // Initialize an empty stack
        $stack = [];

        // Get the length of the number
        $len = strlen($num);

        // Iterate over each digit in the number
        for ($i = 0; $i < $len; $i++) {
            $n = $num[$i];

            // While there are still digits to remove and the top of the stack is greater than the current digit
            while ($k > 0 && count($stack) > 0 && $stack[count($stack) - 1] > $n) {
                // Remove the top digit from the stack
                array_pop($stack);
                // Decrease the count of digits to remove
                $k--;
            }

            // Add the current digit to the stack
            $stack[] = $n;
        }

        // If there are still digits to remove, remove them from the top of the stack
        while ($k > 0) {
            array_pop($stack);
            $k--;
        }

        // Convert the stack to a string
        $res = implode("", $stack);

        // Remove leading zeros
        $res = ltrim($res, "0");

        // If the result is an empty string, return "0", otherwise return the result
        return $res === "" ? "0" : $res;
    }
}
